src/block/Beacon.php

```diff
@@ -23,11 +23,176 @@
+		EffectIds::RESISTANCE => true,
+		EffectIds::SPEED => true,
+		EffectIds::STRENGTH => true
+	];
+
+	private const ALLOWED_SECONDARY_EFFECTS = [
+		EffectIds::REGENERATION => true
+	];
+
+	private int $primaryEffect;
+	private int $secondaryEffect;
+
+	public function readStateFromWorld() : void{
+		parent::readStateFromWorld();
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileBeacon){
+			$this->primaryEffect = $tile->getPrimaryEffect();
+			$this->secondaryEffect = $tile->getSecondaryEffect();
+		}
+	}
+
+	public function writeStateToWorld() : void{
+		parent::writeStateToWorld();
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileBeacon){
+			$tile->setPrimaryEffect($this->primaryEffect);
+			$tile->setSecondaryEffect($this->secondaryEffect);
+		}
+	}
+
+	public function getPrimaryEffect() : int{
+		return $this->primaryEffect;
+	}
+
+	/** @return $this */
+	public function setPrimaryEffect(int $primaryEffect) : self{
+		if(!isset(self::ALLOWED_PRIMARY_EFFECTS[$primaryEffect])){
+			throw new \InvalidArgumentException("Effect ID \"$primaryEffect\" is not allowed in the primary effect");
+		}
+		$this->primaryEffect = $primaryEffect;
+		return $this;
+	}
+
+	public function getSecondaryEffect() : int{
+		return $this->secondaryEffect;
+	}
+
+	/** @return $this */
+	public function setSecondaryEffect(int $secondaryEffect) : self{
+		if(!isset(self::ALLOWED_PRIMARY_EFFECTS[$secondaryEffect]) || !isset(self::ALLOWED_SECONDARY_EFFECTS[$secondaryEffect])){
+			throw new \InvalidArgumentException("Effect ID \"$secondaryEffect\" is not allowed in the secondary effect");
+		}
+		$this->secondaryEffect = $secondaryEffect;
+		return $this;
+	}
 
 	public function getLightLevel() : int{
 		return 15;
 	}
 
-	//TODO
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		if($player instanceof Player){
+			$player->setCurrentWindow(new BeaconInventory($this->position));
+		}
+
+		return true;
+	}
+
+	public function onScheduledUpdate() : void{
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 20 * 3);
+
+		if(!$this->viewSky()){
+			return;
+		}
+
+		$effectIdMap = EffectIdMap::getInstance();
+		$primaryE = $effectIdMap->fromId($this->primaryEffect);
+		$secondaryE = $effectIdMap->fromId($this->secondaryEffect);
+		if($primaryE === null && $secondaryE === null){
+			return;
+		}
+
+		$beaconLevel = 0;
+		for($i = 1; $i <= 4; $i++){
+			if(!$this->isBeaconLevelValid($i)){
+				break;
+			}
+			$beaconLevel++;
+		}
+		if($beaconLevel > 0){
+			$radius = (10 * $beaconLevel) + 10;
+			$effectDuration = 9 + (2 * $beaconLevel);
+
+			$world = $this->position->getWorld();
+			$aabb = (new AxisAlignedBB(0, 0, 0, 1, $world->getMaxY(), 1))->offset($this->position->x, 0, $this->position->z)->expand($radius, 0, $radius);
+			if($primaryE === $secondaryE){
+				if($secondaryE === null){
```

Comment:
This isn't necessary. You already checked if it was null above.

Reply:
For some reason PHPStan doesn’t like it

Reply:
That's because this condition is borked: 
![image](https://user-images.githubusercontent.com/14214667/149838691-9cd65dfb-1922-4036-961a-882827244911.png)

This should only check the primary effect. A beacon may have a primary effect without a secondary effect.